src/fs/rock/RockRebuild.h

```diff
@@ -24,24 +27,81 @@ class LoadingParts;
 
 /// \ingroup Rock
 /// manages store rebuild process: loading meta information from db on disk
-class Rebuild: public AsyncJob
+class Rebuild: public AsyncJob, private IndependentRunner
 {
     CBDATA_CHILD(Rebuild);
 
+private:
+    /// low-level anti-padding storage class for LoadingEntry and LoadingSlot flags
+    class LoadingFlags
+    {
+    public:
+        LoadingFlags(): state(0), anchored(0), mapped(0), finalized(0), freed(0) {}
+
+        /* for LoadingEntry */
+        uint8_t state:3;  ///< current entry state (one of the LoadingEntry::State values)
+        uint8_t anchored:1;  ///< whether we loaded the inode slot for this entry
+
+        /* for LoadingSlot */
+        uint8_t mapped:1;  ///< whether the slot was added to a mapped entry
+        uint8_t finalized:1;  ///< whether finalizeOrThrow() has scanned the slot
+        uint8_t freed:1;  ///< whether the slot was given to the map as free space
+    };
+
+    friend class LoadingEntry;
+    friend class LoadingSlot;
+
 public:
     Rebuild(SwapDir *dir);
     virtual ~Rebuild() override;
 
+    /// shared metadata during rebuild process
+    class Metadata
+    {
+    public:
+        size_t sharedMemorySize() const { return sizeof(*this); }
+        static size_t SharedMemorySize() { return sizeof(Metadata); }
+
+        StoreRebuildData counts;
+    };
+
+    typedef Ipc::StoreMapItems<uint64_t> Sizes;
+    typedef Ipc::StoreMapItems<uint32_t> Versions;
+    typedef Ipc::StoreMapItems<Ipc::StoreMapSliceId> Mores;
+    typedef Ipc::StoreMapItems<LoadingFlags> Flags;
+
+    static Ipc::Mem::Owner<Metadata> *InitMetadata(const SwapDir *dir);
+
+    /// whether the current kid is responsible for rebuilding this db file
+    static bool ShouldStart(const SwapDir &);
+
+    /* Registered Runner API */
+    virtual void startShutdown() override;
+
 protected:
     /* AsyncJob API */
     virtual void start() override;
     virtual bool doneAll() const override;
     virtual void swanSong() override;
 
+    bool loadedAndValidated() const { return doneLoading() && doneValidating(); }
     bool doneLoading() const;
     bool doneValidating() const;
 
 private:
+    class LoadingPartsOwner
```

Comment:
Please try to move this and related declarations back to src/fs/rock/RockRebuild.cc. I hope that forward-declaring these internal classes in the header is sufficient.